include/PetriEngine/Colored/VariableVisitor.h

```diff
@@ -289,6 +295,32 @@ namespace PetriEngine {
 
                 get_variables(e, variables, varPositions, varModifierMap, tuples, false);
             }
+
+        private:
+            bool is_last_variable_expr(const std::shared_ptr<const ColorExpression> expr) const {
+                std::stack<std::shared_ptr<const ColorExpression>> stack;
```

Comment:
Again, I don't think you need to have shared pointers. These will not go out of their lifecycle while you work on them here.

Reply:
(It is not free to move smart-pointers around, so it is a good habit to think about if they realy are needed -- even if the impact is minor)